File: alienDictionary/aliendict.py
from collections import defaultdict, Counter, deque

def alienOrder( words):
    
    # Step 0: create data structures + the in_degree of each unique letter to 0.
    adj_list = defaultdict(set)
    print("adj_list",adj_list)
    in_degree = Counter({c : 0 for word in words for c in word})
    print("in_degree",in_degree)
    # Step 1: We need to populate adj_list and in_degree.
    # For each pair of adjacent words...
    print("zip 13",words, words[1:])
    for first_word, second_word in zip(words, words[1:]):
        print("first_word",first_word)
        print("second_word",second_word)
        for c, d in zip(first_word, second_word):
            
            print("c",c)
            print("d",d)
            if c != d:
                if d not in adj_list[c]:
                    adj_list[c].add(d)
                    in_degree[d] += 1
                break
        else: # Check that second word isn't a prefix of first word.
            if len(second_word) < len(first_word): return ""
    print(adj_list,'adj_list')
    # Step 2: We need to repeatedly pick off nodes with an indegree of 0.
    output = []
    print("indgree",in_degree)
    queue = deque([c for c in in_degree if in_degree[c] == 0])
    
    print("queue",queue)
    while queue:
        c = queue.popleft()
        output.append(c)
        for d in adj_list[c]:
            in_degree[d] -= 1
            if in_degree[d] == 0:
                queue.append(d)
                
    # If not all letters are in output, that means there was a cycle and so
    # no valid ordering. Return "" as per the problem description.
    if len(output) < len(in_degree):
        return ""
    # Otherwise, convert the ordering we found into a string and return it.
    return "".join(output)
print(alienOrder(["wrt", "wrf", "er", "ett", "rftt"]))

File: alienDictionary/alienDict.js
function alienOrder(words) {
  const adjList = {};
  const inDegree = words.reduce((acc, params) => {
    // console.log({ params });
    for (char of params) {
      if (!acc[char]) acc[char] = 0;
    }
    return acc;
  }, {});
  const shiftWordsByOne = [...words];
  shiftWordsByOne.shift();
  console.log({ words, shiftWordsByOne });
  for (let i = 0; i < shiftWordsByOne.length; i++) {
    const first = words[i];
    const second = shiftWordsByOne[i];
    console.log({ first, second });
    let lengthRef = Math.min(first.length, second.length);
    console.log({ lengthRef });
    for (let i = 0; i < lengthRef; i++) {
      let c = first[i];
      let d = second[i];
      console.log({ c, d });
      if (c !== d) {
        if (!adjList[c]) {
          adjList[c] = new Set();
          if (!adjList[c].has(d)) {
            adjList[c].add(d);
            inDegree[d] += 1;
          }
        }
        break;
      } else {
        if (second.length < first.length) return "";
      }
    }
  }
  console.log(adjList);
  console.log(inDegree);
  const output = [];
  const queue = [];
  for (let degree in inDegree) {
    if (inDegree[degree] === 0) queue.push(degree);
  }
  console.log({ queue });
  while (queue.length > 0) {
    let c = queue.shift();
    console.log({ c });
    output.push(c);
    console.log({ output });
    console.log(adjList, c, "test adjustList");
    for (let d of adjList[c]) {
      console.log("test", { d });
      inDegree[d] -= 1;
      if (inDegree[d] === 0) {
        queue.push(d);
      }
    }
  }
}

alienOrder(["wrt", "wrf", "er", "ett", "rftt"]);
const words2 = ["wrt", "wrf", "er", "ett", "rftt"];
console.log(words2);
words2.shift();
console.log(words2);
